spec/index.bs

```diff
@@ -6518,9 +6518,102 @@ and link it together with the shaders provided by the user.
 This linking is one of the reason the object is created as a whole.
 
 This combination state is created as a single object
-(by {{GPUDevice/createComputePipeline(descriptor)|GPUDevice.createComputePipeline()}} or {{GPUDevice/createRenderPipeline(descriptor)|GPUDevice.createRenderPipeline()}}),
-and switched as one
-(by {{GPUComputePassEncoder/setPipeline(pipeline)|GPUComputePassEncoder.setPipeline}} or {{GPURenderCommandsMixin/setPipeline(pipeline)|GPURenderCommandsMixin.setPipeline}} correspondingly).
+(a {{GPUComputePipeline}} or {{GPURenderPipeline}})
+and switched using one command
+({{GPUComputePassEncoder}}.{{GPUComputePassEncoder/setPipeline()}} or
+{{GPURenderCommandsMixin}}.{{GPURenderCommandsMixin/setPipeline()}} respectively).
+
+There are two ways to create pipelines:
+
+: <dfn dfn>immediate pipeline creation</dfn>
+:: {{GPUDevice/createComputePipeline()}} and {{GPUDevice/createRenderPipeline()}}
+    return a pipeline object which can be used immediately in a pass encoder.
+
+    When this fails, the pipeline object will be invalid and the call will generate either a
+    [$validation error$] or an [$internal error$].
+
+    Note:
+    A handle object is returned immediately, but actual pipeline creation is not synchronous.
+    If pipeline creation takes a long time, this can incur a stall in the
+    [=device timeline=] at some point between the creation call and execution of the
+    {{GPUQueue/submit()}} in which it is first used.
+    The point is unspecified, but most likely to be one of: at creation, at the first usage of the
+    pipeline in `setPipeline()`, at the corresponding `finish()` of that {{GPUCommandEncoder}} or
+    {{GPURenderBundleEncoder}}, or at {{GPUQueue/submit()}} of that {{GPUCommandBuffer}}.
+
+: <dfn dfn>async pipeline creation</dfn>
+:: {{GPUDevice/createComputePipelineAsync()}} and {{GPUDevice/createRenderPipelineAsync()}}
+    return a `Promise` which resolves to a pipeline object when creation of the pipeline has
+    completed.
+
+    When this fails, the `Promise` rejects with a {{GPUPipelineError}}.
+
+<dfn interface>GPUPipelineError</dfn> describes a pipeline creation failure.
+
+Issue: According to [whatwg/webidl#1211](https://github.com/whatwg/webidl/pull/1211),
+to conform to the standard pattern, `message` should be optional and default to `""`.
+
+<script type=idl>
+[Exposed=(Window, DedicatedWorker), SecureContext, Serializable]
+interface GPUPipelineError : DOMException {
+    constructor(DOMString message, GPUPipelineErrorInit options);
+    readonly attribute GPUPipelineErrorReason reason;
+};
+
+dictionary GPUPipelineErrorInit {
+    required GPUPipelineErrorReason reason;
+};
+
+enum GPUPipelineErrorReason {
+    "validation",
+    "internal"
```

Comment:
This PR added a new GPUPipelineError, which can be either of reason "validation", or "internal"; but then we also have existing `GPUValidationError` and `GPUInternalError` error types?

Is there a rationale for having this seeming duplication or two-layer aspect to the errors? If a pipeline creation fails, could it not raise one of the existing GPUValidationError or GPUInternalError types of errors?

Reply:
Ping @kainino0x , I am still puzzling over this a little bit?

Reply:
Sorry, have been temporarily unavailable on account of a certain situation at my employer. Anyway:

Yes, there's a reason for this. Mentioned above, the web spec Promises guide says promises must only reject with `Error` instances. The `GPUError` types, perhaps poorly named, don't inherit from `Error` because they're not synchronous errors with JS stacks (and they're things we `return`, not things we `throw`/reject with).

Additionally, the main reason for this change in the first place was that we wanted to make sure we had a place to provide rich info back to the caller as we would with the non-async versions. The non-async versions return a pipeline object, say `GPURenderPipeline`, and we would be able to put structured messages on there the same way we do now with `GPUShaderModule.compilationInfo()`, something like `GPURenderPipeline.creationInfo()`. The async versions reject, but we would want to be able to return it in the same form, e.g. `GPUPipelineError.creationInfo()`.

As for why we have an enum instead of an inheritance tree, I guess it's just that we don't expect to add any new types or shapes of GPUPipelineErrors in the future, and the structure of that "creationInfo" data is expected to be generic enough for both internal and validation errors.

Reply:
Gotcha, thanks!